src/server/streamableHttp.ts

```diff
@@ -0,0 +1,368 @@
+   */
+  sessionId: string | undefined;
+
+  /**
+   * Custom headers to be included in all responses
+   * These headers will be added to both SSE and regular HTTP responses
+   */
+  customHeaders?: Record<string, string>;
+
+}
+
+/**
+ * Server transport for Streamable HTTP: this implements the MCP Streamable HTTP transport specification.
+ * It supports both SSE streaming and direct HTTP responses.
+ * 
+ * Usage example:
+ * 
+ * ```typescript
+ * // Stateful mode - server sets the session ID
+ * const statefulTransport = new StreamableHTTPServerTransport({
+ *  sessionId: randomUUID(),
+ * });
+ * 
+ * // Stateless mode - explicitly set session ID to undefined
+ * const statelessTransport = new StreamableHTTPServerTransport({
+ *    sessionId: undefined,
+ * });
+ * 
+ * // Using with pre-parsed request body
+ * app.post('/mcp', (req, res) => {
+ *   transport.handleRequest(req, res, req.body);
+ * });
+ * ```
+ * 
+ * In stateful mode:
+ * - Session ID is generated and included in response headers
+ * - Session ID is always included in initialization responses
+ * - Requests with invalid session IDs are rejected with 404 Not Found
+ * - Non-initialization requests without a session ID are rejected with 400 Bad Request
+ * - State is maintained in-memory (connections, message history)
+ * 
+ * In stateless mode:
+ * - Session ID is only included in initialization responses
+ * - No session validation is performed
+ */
+export class StreamableHTTPServerTransport implements Transport {
+  // when sessionId is not set (undefined), it means the transport is in stateless mode
+  private _sessionId: string | undefined;
+  private _started: boolean = false;
+  private _customHeaders: Record<string, string>;
+  private _sseResponseMapping: Map<RequestId, ServerResponse> = new Map();
+
+  onclose?: () => void;
+  onerror?: (error: Error) => void;
+  onmessage?: (message: JSONRPCMessage) => void;
+
+  constructor(options: StreamableHTTPServerTransportOptions) {
+    this._sessionId = options.sessionId;
+    this._customHeaders = options.customHeaders || {};
+  }
+
+  /**
+   * Starts the transport. This is required by the Transport interface but is a no-op
+   * for the Streamable HTTP transport as connections are managed per-request.
+   */
+  async start(): Promise<void> {
+    if (this._started) {
+      throw new Error("Transport already started");
+    }
+    this._started = true;
+  }
+
+  /**
+   * Handles an incoming HTTP request, whether GET or POST
+   */
+  async handleRequest(req: IncomingMessage, res: ServerResponse, parsedBody?: unknown): Promise<void> {
+    if (req.method === "GET") {
+      await this.handleGetRequest(req, res);
+    } else if (req.method === "POST") {
+      await this.handlePostRequest(req, res, parsedBody);
+    } else if (req.method === "DELETE") {
+      await this.handleDeleteRequest(req, res);
+    } else {
+      res.writeHead(405, this._customHeaders).end(JSON.stringify({
```

Comment:
This response is missing an `Allow` header (required in HTTP for 405s). Should we just unify this codepath with `handleGetRequest`?